### PR TITLE
Added retain_graph as a kwarg to the main engine backward function

### LINKED ISSUE #329: Difference between loss.backward() and model_engine.backward(loss) ?
Hi , 

I have tried out both loss.backward() and model_engine.backward(loss) for my code. There are several subtle differences that I have observed , for one retain_graph = True does not work for model_engine.backward(loss) . This is creating a problem since buffers are not being retained every time I run the code for some reason. 

Please look into this if you could.

### PR DESCRIPTION
**Pull request by ncilfone**
Added retain_graph as a kwarg to the main engine backward function -- this allows deepspeed's backward call to support multiple calls to backward with multiple losses since retain_graph is needed for the first backward call.

Closes #329

### CODE REVIEW

**Review by jeffra — APPROVED** [review]

**Review by raghavthind2005 — APPROVED** [review]

**Comment by rocm-mici** [comment]
Can one of the admins verify this patch?

**Comment by buttercutter** [comment]
@ncilfone @tjruwase  May I ask what exact kind of branch conflicts for this pull request ?

**Comment by tjruwase** [comment]
@buttercutter , I have resolved the merge conflict. 

@ncilfone, FYI and apologies for delay on this.
